### DISCUSSION TITLE
Providing Credentials to Gradle

### DISCUSSION THREAD

**Discussion by knechtionscoding** [renovatebot/renovate#19363] [Request Help] [answered]
I'm attempting to provide credentials to allow renovate to look up gradle dependencies in a private artifactory.

I've got docker credentials configured and working for the same host.

```json
"packageRules": [
     ...
     {
            matchHost: 'https://<artifactory hostname>/',
            username: "{{ secrets.username }}",
            password: "{{ secrets.password }}",
          }
        ],
       ...
      "secrets": {
          "username": process.env.DOCKER_USERNAME,
          "password": process.env.DOCKER_PASSWORD,
        }
```

Still telling me it can't look up the dependencies properly. I know the credentials are valid because the docker part works correctly and docker dependencies are being pulled from the private repository

**Comment by rarkins** [comment, marked as answer]
How would you normally pass credentials to gradle, e.g. in CI pipelines?

**Reply by knechtionscoding** [reply]
Something like:

```
    contextUrl = "<url>
    resolve {
        repository {
            repoKey = "maven"
            username = System.env.ARTIFACTORY_USERNAME
            password = System.env.ARTIFACTORY_PASSWORD
            maven = true
        }
    }
```

In the build.gradle. Do I simply need to set the same env variables and when renovate runs it'll use gradle directly? Or does renovate reimplement gradle such that build.gradle isn't relevant other than regex?

Based on: https://docs.renovatebot.com/modules/manager/gradle/#additional-information I figured that I needed to provide a host rule for it to find dependencies...

**Reply by rarkins** [reply]
hostRules is definitely the way to configure Renovate. I was trying to work out how we could subsequently pass that through to Gradle. Are you self hosting?

**Reply by knechtionscoding** [reply]
Yes, I am self-hosting

**Reply by knechtionscoding** [reply]
If there was a general method for me to provide credentials. I know it can authenticate to the private repo (do it for docker and npm). It's just gradle that doesn't do it.

**Reply by rarkins** [reply]
In that case, you can do it similar to what you did before, but you just need to make sure that Renovate is passing through the env variables you need when it calls Gradle.

Simplest way to do that is `exposeAllEnv=true`. If you don't want to expose all env to Gradle, then you can configure `customEnvVariables` individually.

**Reply by knechtionscoding** [reply]
Okay, I'll give that a try. I may also try skipping the secret part and just env variables as described in the docs under Gradle (via maven). I'll try both methods and see what works

**Comment by knechtionscoding** [comment]
I set both exposeAllEnv and configured the secrets properly, realized I was mixing config files. It appears to be pulling dependencies properly.

**Comment by fisher-zwl** [comment]
你好!你的邮件已收到，谢谢！
